File: bootstrap/src/main/java/net/dontdrinkandroot/wicket/bootstrap/css/grid/ColumnOffsetStack.java
/*
 * Copyright (C) 2012-2017 Philip Washington Sorst <philip@sorst.net>
 * and individual contributors as indicated
 * by the @authors tag.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.dontdrinkandroot.wicket.bootstrap.css.grid;

public class ColumnOffsetStack implements ColumnOffset
{
    private ColumnOffsetDefault columnOffsetDefault;

    private ColumnOffsetSmall columnOffsetSmall;

    private ColumnOffsetMedium columnOffsetMedium;

    private ColumnOffsetLarge columnOffsetLarge;

    public ColumnOffsetStack(
            ColumnOffsetDefault columnOffsetDefault,
            ColumnOffsetSmall columnOffsetSmall,
            ColumnOffsetMedium columnOffsetMedium,
            ColumnOffsetLarge columnOffsetLarge
    )
    {
        super();
        this.columnOffsetDefault = columnOffsetDefault;
        this.columnOffsetSmall = columnOffsetSmall;
        this.columnOffsetMedium = columnOffsetMedium;
        this.columnOffsetLarge = columnOffsetLarge;
    }

    @Override
    public String getClassString()
    {
        StringBuffer classBuffer = new StringBuffer();
        if (null != this.columnOffsetDefault) {
            classBuffer.append(this.columnOffsetDefault.getClassString());
            classBuffer.append(" ");
        }
        if (null != this.columnOffsetSmall) {
            classBuffer.append(this.columnOffsetSmall.getClassString());
            classBuffer.append(" ");
        }
        if (null != this.columnOffsetMedium) {
            classBuffer.append(this.columnOffsetMedium.getClassString());
            classBuffer.append(" ");
        }
        if (null != this.columnOffsetLarge) {
            classBuffer.append(this.columnOffsetLarge.getClassString());
            classBuffer.append(" ");
        }

        return classBuffer.toString().trim();
    }

    @Override
    public ColumnSize getInverseColumnSize()
    {
        ColumnSizeDefault columnSizeDefault =
                this.columnOffsetDefault == null ? null : this.columnOffsetDefault.getInverseColumnSize();
        ColumnSizeSmall columnSizeSmall =
                this.columnOffsetSmall == null ? null : this.columnOffsetSmall.getInverseColumnSize();
        ColumnSizeMedium columnSizeMedium =
                this.columnOffsetMedium == null ? null : this.columnOffsetMedium.getInverseColumnSize();
        ColumnSizeLarge columnSizeLarge =
                this.columnOffsetLarge == null ? null : this.columnOffsetLarge.getInverseColumnSize();

        return new ColumnSizeStack(columnSizeDefault, columnSizeSmall, columnSizeMedium, columnSizeLarge);
    }

    @Override
    public ColumnOffset getInverseColumnOffset()
    {
        ColumnOffsetDefault columnOffsetDefault =
                this.columnOffsetDefault == null ? null : this.columnOffsetDefault.getInverseColumnOffset();
        ColumnOffsetSmall columnOffsetSmall =
                this.columnOffsetSmall == null ? null : this.columnOffsetSmall.getInverseColumnOffset();
        ColumnOffsetMedium columnOffsetMedium =
                this.columnOffsetMedium == null ? null : this.columnOffsetMedium.getInverseColumnOffset();
        ColumnOffsetLarge columnOffsetLarge =
                this.columnOffsetLarge == null ? null : this.columnOffsetLarge.getInverseColumnOffset();

        return new ColumnOffsetStack(columnOffsetDefault, columnOffsetSmall, columnOffsetMedium, columnOffsetLarge);
    }
}

File: bootstrap/src/main/kotlin/net/dontdrinkandroot/wicket/bootstrap/css/grid/ColumnOffsetStack.kt
package net.dontdrinkandroot.wicket.bootstrap.css.grid

class ColumnOffsetStack(
    private val columnOffsetDefault: ColumnOffset?,
    private val columnOffsetSmall: ColumnOffset?,
    private val columnOffsetMedium: ColumnOffset?,
    private val columnOffsetLarge: ColumnOffset?
) : ColumnOffset {

    override val classString: String
        get() {
            val classBuffer = StringBuffer()
            if (null != columnOffsetDefault) {
                classBuffer.append(columnOffsetDefault.classString)
                classBuffer.append(" ")
            }
            if (null != columnOffsetSmall) {
                classBuffer.append(columnOffsetSmall.classString)
                classBuffer.append(" ")
            }
            if (null != columnOffsetMedium) {
                classBuffer.append(columnOffsetMedium.classString)
                classBuffer.append(" ")
            }
            if (null != columnOffsetLarge) {
                classBuffer.append(columnOffsetLarge.classString)
                classBuffer.append(" ")
            }

            return classBuffer.toString().trim { it <= ' ' }
        }

    override val inverseColumnSize: ColumnSize
        get() {
            val columnSizeDefault = if (columnOffsetDefault == null) null else columnOffsetDefault.inverseColumnSize
            val columnSizeSmall = if (columnOffsetSmall == null) null else columnOffsetSmall.inverseColumnSize
            val columnSizeMedium = if (columnOffsetMedium == null) null else columnOffsetMedium.inverseColumnSize
            val columnSizeLarge = if (columnOffsetLarge == null) null else columnOffsetLarge.inverseColumnSize
            return ColumnSizeStack(columnSizeDefault, columnSizeSmall, columnSizeMedium, columnSizeLarge)
        }

    override val inverseColumnOffset: ColumnOffset
        get() {
            val columnOffsetDefault = if (columnOffsetDefault == null) null else columnOffsetDefault.inverseColumnOffset
            val columnOffsetSmall = if (columnOffsetSmall == null) null else columnOffsetSmall.inverseColumnOffset
            val columnOffsetMedium = if (columnOffsetMedium == null) null else columnOffsetMedium.inverseColumnOffset
            val columnOffsetLarge = if (columnOffsetLarge == null) null else columnOffsetLarge.inverseColumnOffset
            return ColumnOffsetStack(columnOffsetDefault, columnOffsetSmall, columnOffsetMedium, columnOffsetLarge)
        }
}